crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -36,26 +38,85 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
         query: String,
         operation: Option<String>,
         options: QueryPlanOptions,
-    ) -> Result<Arc<QueryPlan>, QueryPlannerError> {
-        if let Some(value) = self
-            .cached
-            .lock()
-            .await
-            .get(&(query.clone(), operation.clone(), options.clone()))
-            .cloned()
-        {
+    ) -> PlanResult {
+        let mut locked_cache = self.cached.lock().await;
+        let key = (query, operation, options);
+        if let Some(value) = locked_cache.get(&key).cloned() {
             return value;
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        // We must only drop the locked cache after we have locked the
+        // wait map. Otherwise,we might get a race that causes us to
+        // miss a broadcast.
+        drop(locked_cache);
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                // Our use case is very specific, so we are sure
+                // that we won't get any errors here.
+                let (recv_key, recv_plan) = receiver.recv().await.expect("recv() cannot fail here");
```

Comment:
You can actually put the proof message in the expect.

```suggestion
                let (recv_key, recv_plan) = receiver.recv().await
                    .expect("the sender won't ever be dropped before all the receivers finish; qed");
```

Reply:
done